src/realm/alloc.hpp

```diff
@@ -35,6 +35,10 @@ class Replication;
 
 using ref_type = size_t;
 
+// Maximum .realm file size and also maximum in-memory transaction snapshot size. On 32-bit architectures we need 
+// to cap it to 2^31 * 0.80 bytes because arithmetic on references can otherwise overflow in various places.
+constexpr size_t maximum_realm_size = static_cast<size_t>((std::numeric_limits<ref_type>::max() >> 1) * 0.8);
```

Comment:
I don't think that this solution conforms with the usual precision of our designs. For one, 80% is an arbitrary limit which could now fail previously working (albeit poorly designed) apps. Additionally, it doesn't guarantee to actually solve all of these assertions in cases where the size multiplies.

I would prefer that we check all our arithmetic for overflows individually case by case and throw where appropriate.